Add documentation page for checks and assertions (#33241)
hashicorp/terraform: website/docs/language/checks/index.mdx

  }
}
```

### Scoped Data Sources

You can use any data source from any provider as a scoped data source within a `check` block.

A `check` block can optionally contain a nested (a.k.a scoped) data source. This `data` block behaves like an external [data source](/terraform/language/data-sources), except you can not reference it outside its enclosing `check` block. Additionally, if a scoped data source's provider raises any errors, they are masked as warnings and do not prevent Terraform from continuing operation execution.

You can use a scoped data source to validate the status of a piece of infrastructure outside of the usual Terraform resource lifecycle. [In the above example](#checks-syntax), if the `terraform_io` data source fails to load, you receive a warning instead of a blocking error, which would occur if you declared this data source outside of a `check` block.

#### Meta-Arguments

Scoped data sources support the `depends_on` and `provider` [meta-arguments](/terraform/language/resources/syntax#meta-arguments). Scoped data sources do not support the `count` or`for_each` meta-arguments.

> ##### `depends_on`

Author: @omarismail @rkoron007 This is a new section, so in need of fresh review.

The `depends_on` meta argument can be particularly powerful when used within scoped data sources.

> The [previous example](#checks-syntax) would fail during any initial planning, as the configuration hasn't been applied so the website simply would not exist. As written, this would be expected behaviour but is not particularly useful and is potentially noisy as other genuine errors in the configuration could be obscured by the warning message.

Reviewer: nit: but i would italitcs _initial_, as we want to emphasize the fact that the prior example **only** would fail during the first time cause the resoruce isn't created yet

An addition of the `depends_on` attribute, linking the scoped data source to the highest-levelled piece of infrastructure such as a load balancer, would mean that during any initial planning the check would instead report as `known after apply`. In addition, it would not produce any unnecessary or noisy warnings. Going forward, the check would execute again for both plans and applies providing the required validation.

> One caveat to the `depends_on` attribute is that any changes to the dependencies would again cause the check block to report as `known after apply`, essentially meaning the check block would not execute during that planning stage. Depending on your use case this behavior could be acceptable or problematic.

Reviewer 2: I tried to make this clear and re-write in active voice, but please feel free to push back if this is no longer technically accurate:
```suggestion
One problem with this strategy is that if the resource your scoped data source `depends_on` changes, the check block returns `known after apply` until Terraform has recreated that resource. Depending on your use case, this behavior could be acceptable or problematic.
```

> 

Reviewer: I think we should add an extra paragraph to state our guidance, such as when writing a check that depends on a resource being created, best to use depends_on

Author: Done!

Condition arguments within `assert` blocks can refer to scoped data sources within the enclosing `check` block and any variables, resources, data sources, or module outputs within the current module.

[Learn more about assertions](/terraform/language/expressions/custom-conditions#check-assertions).

### Meta-Arguments

Check blocks do not currently support [meta-arguments](/terraform/language/resources/syntax#meta-arguments). We are still collecting feedback on this feature, so if your use case would benefit from check blocks supporting meta-arguments, please [let us know](https://github.com/hashicorp/terraform/issues/new/choose).

## Continuous Validation in Terraform Cloud

Terraform Cloud can automatically validate whether checks in a workspace’s configuration continue to pass after Terraform provisions new infrastructure. See [Continuous Validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for details.

> ## Choosing Checks or other Custom Conditions

Author: @rkoron007, I've added a lot to this section. I think it's worth a fresh review. Just from here to the end of the file.

> Check blocks offer the most flexible validation solution within Terraform. As you can reference outputs, variables, resources and data sources within check assertions, they can be used to model all the alternate [Custom Conditions](/terraform/language/expressions/custom-conditions). However, that does not mean you should replace all your custom conditions with check blocks.

Reviewer 2: This is such a big point, I can't help but want to add a _little_ bit of emphasis. 
```suggestion
Check blocks offer the most _flexible_ validation solution within Terraform. You can reference outputs, variables, resources, and data sources within check assertions. You can also use checks to model every alternate [Custom Conditions](/terraform/language/expressions/custom-conditions). However, that does not mean you should replace all your custom conditions with check blocks.
```

> As discussed, the major behavioral difference between [check block assertions](#assertions) and other custom conditions is that they do not affect Terraform's execution of an operation. It is this behaviour you should think about when deciding which validation to apply.

Reviewer 2: ```suggestion
There are major behavioral differences between check block assertions and other custom conditions, the main one being that check blocks do _not_ affect Terraform's execution of an operation. You can use this non-blocking behavior to decide the best type of validation for your use case.
```

### Outputs and Variables

[Output postconditions](/terraform/language/expressions/custom-conditions#outputs) and [variable validations](/terraform/language/expressions/custom-conditions#input-variable-validation) both make assertions around inputs and outputs.

> In these cases it is often desirable that Terraform should block further execution. For example, it is not useful to warn that an input variable was invalid after an entire Terraform configuration has been applied based on that invalid value. This is the behaviour a check block would apply if an assertion was made against a variable. A validation block against the variable itself would alert before any configuration was applied that the provided input was invalid, and then not create infrastructure based on a flawed premise.

Reviewer 2: I like the flow of splitting this up into the main point, then the example explaining why. I also rephrased a bit for clarity and active voice.  
```suggestion
In some cases, you might want Terraform to block further execution. 

For example, it is not helpful for Terraform to warn that an input variable is invalid after it applies an entire configuration with that input variable. In this case, a check block would warn of the invalid input variable _without_ interrupting the operation. A validation block for the same input variable would alert you of the invalid variable _before_ applying the configuration.
```

> ### Resource preconditions and postconditions

Reviewer 2: Sentence case:
```suggestion
### Resource Preconditions and Postconditions
```

> The boundary between resource and data source [preconditions and postconditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) and check blocks is more nuanced.

Reviewer 2: For brevity:
```suggestion
The difference between [preconditions and postconditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) and check blocks is more nuanced.
```

> Preconditions are unique amongst the custom conditions in that they execute before a resource change is applied or planned, rather than after. This is discussed in detail in [Choosing Between Preconditions and Postconditions](/terraform/language/expressions/custom-conditions#choosing-between-preconditions-and-postconditions) and the same advice applies when choosing between a precondition and a check block.

Reviewer 2: Rewrote to avoid passive voice: 
```suggestion
Preconditions are unique amongst the custom conditions in that they execute _before_ a resource change is applied or planned. [Choosing Between Preconditions and Postconditions](/terraform/language/expressions/custom-conditions#choosing-between-preconditions-and-postconditions) offers guidance on choosing between a precondition and a postcondition, and the same topics also apply to choosing between preconditions and a check block.
```

> Postconditions, however, can often be used interchangeably with check blocks to validate resources and data sources.

Reviewer 2: ```suggestion
You can often use postconditions interchangeably with check blocks to validate resources and data sources.
```

For example, you can [rewrite the above `check` block example](#checks-syntax) to use a postcondition instead. The below code uses a `postcondition` block to validate that the Terraform website returns the expected status code of `200`.

```hcl
data "http" "terraform_io" {
  url = "https://www.terraform.io"

  lifecycle {
    postcondition {
        condition = self.status_code == 200
        error_message = "${self.url} returned an unhealthy status code"
    }
  }
}
```

> Both of our `check` and `postcondition` block examples validate that the Terraform website returns a `200` status code during a plan or an apply operation. The difference between the two blocks is how each handles failure.

Reviewer 2: ```suggestion
Both the `check` and `postcondition` block examples validate that the Terraform website returns a `200` status code during a plan or an apply operation. The difference between the two blocks is how each handles failure.
```

> If a `postcondition` block fails, it _blocks_ Terraform from continuing execution of the current operation. If a `check` block fails, it _does not_ block Terraform's execution of an operation.

Reviewer 2: ```suggestion
If a `postcondition` block fails, it _blocks_ Terraform from executing the current operation. If a `check` block fails, it _does not_ block Terraform from executing an operation.
```

Condition arguments within `assert` blocks can refer to scoped data sources within the enclosing `check` block and any variables, resources, data sources, or module outputs within the current module.

[Learn more about assertions](/terraform/language/expressions/custom-conditions#check-assertions).

### Meta-Arguments

Check blocks do not currently support [meta-arguments](/terraform/language/resources/syntax#meta-arguments). We are still collecting feedback on this feature, so if your use case would benefit from check blocks supporting meta-arguments, please [let us know](https://github.com/hashicorp/terraform/issues/new/choose).

## Continuous Validation in Terraform Cloud

Terraform Cloud can automatically validate whether checks in a workspace’s configuration continue to pass after Terraform provisions new infrastructure. See [Continuous Validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for details.

## Choosing Checks or other Custom Conditions

> Check blocks offer the most _flexible_ validation solution within Terraform. You can reference outputs, variables, resources, and data sources within check assertions. You can also use checks to model every alternate [Custom Conditions](/terraform/language/expressions/custom-conditions). However, that does not mean you should replace all your custom conditions with check blocks.

Reviewer 2: ```suggestion
Check blocks offer the most _flexible_ validation solution within Terraform. You can reference outputs, variables, resources, and data sources within check assertions. You can also use checks to model every alternate [Custom Condition](/terraform/language/expressions/custom-conditions). However, that does not mean you should replace all your custom conditions with check blocks.
```

There are major behavioral differences between check block assertions and other custom conditions, the main one being that check blocks do _not_ affect Terraform's execution of an operation. You can use this non-blocking behavior to decide the best type of validation for your use case.

### Outputs and Variables

[Output postconditions](/terraform/language/expressions/custom-conditions#outputs) and [variable validations](/terraform/language/expressions/custom-conditions#input-variable-validation) both make assertions around inputs and outputs.

> In some cases, you might want Terraform to block further execution. 

Reviewer 2: I think we can make this sentence clearer with stronger wording, like so:
```suggestion
This is one of the cases where you might want Terraform to block further execution. 
```

For example, it is not helpful for Terraform to warn that an input variable is invalid after it applies an entire configuration with that input variable. In this case, a check block would warn of the invalid input variable _without_ interrupting the operation. A validation block for the same input variable would alert you of the invalid variable and halt the plan or apply operation.

### Resource Preconditions and Postconditions

The difference between [preconditions and postconditions](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) and check blocks is more nuanced.

> Preconditions are unique amongst the custom conditions in that they execute _before_ a resource change is applied or planned. [Choosing Between Preconditions and Postconditions](/terraform/language/expressions/custom-conditions#choosing-between-preconditions-and-postconditions) offers guidance on choosing between a precondition and a postcondition, and the same topics also apply to choosing between preconditions and a check block.

Reviewer 2: Tiny tweak for grammar (this was my bad):
```suggestion
Preconditions are unique amongst the custom conditions in that they execute _before_ a resource change is applied or planned. [Choosing Between Preconditions and Postconditions](/terraform/language/expressions/custom-conditions#choosing-between-preconditions-and-postconditions) offers guidance on choosing between a precondition and a postcondition, and the same topics also apply to choosing between a precondition and a check block.
```

---

> page_title: Checks - Configuration Language

Author: My thinking here is that at this level the other pages refer to things like Resources, Data Sources, and Providers even though within the syntax these are actually Resource blocks, Data source blocks, provider blocks, etc. I thought that if we titled this page Check blocks it wouldn't match the style.

I don't really mind what we call it, I just wanted to give my thought process. FWIW Custom Condition Checks is buried within a submodule so they don't share the same space on the page if they makes a difference.

description: >-
  Check custom requirements for infrastructure to provide ongoing and continuous validation.
---

# Checks

-> **Note:** Check blocks are only available in Terraform v1.5.0 and later.

> You can use `check` blocks to validate your infrastructure outside the usual resource lifecycle. Check blocks allow you to define [Custom Conditions](/terraform/language/expressions/custom-conditions) that execute on every Terraform plan or apply operation without affecting the overall status of the operation.

Reviewer: We should add:
> Checks execute as the last step of a plan/apply, after the infrastructure has been planned/provisioned.

Author: Done!

```hcl
check "health_check" {
  data "http" "terraform_io" {
    url = "https://www.terraform.io"
  }

  assert {
    condition = data.http.terraform_io.status_code == 200
    error_message = "${data.http.terraform_io.url} returned an unhealthy status code"
  }
}
```

### Scoped Data Sources

A `check` block can optionally have a nested data source. This block behaves in the same way as normal/external [data sources](/terraform/language/data-sources), except it cannot be referenced from outside the enclosing check block. In addition, any errors or failures raised by the provider of the data source will be masked as warnings within the output and will not prevent the continued execution of the overall Terraform operation.

A scoped data source allows you to validate the status of a piece of infrastructure outside the usual Terraform resource lifecycle. In the example in the previous section, if the `terraform_io` data source fails to load we would simply receive a warning that the `http` block couldn't be loaded rather than the blocking error that would be raised if the data source was defined outside a check block.

> Any data source from any provider can be used within a check block.

Author: I moved this first, and changed it to: "You can use any data source from any provider as a scoped data source within a `check` block."

Since it's now the first line, highlighting the scoped data source seemed important.

### Assertions

Each `check` block must have at least one, but potentially many, `assert` blocks. An `assert` block has a `condition` attribute and an `error_message` attribute, that match the definitions within [Custom Conditions](/terraform/language/expressions/custom-conditions).

Unlike other custom conditions, assertions do not affect the execution of the overall Terraform operation. A failed assertion will report a warning diagnostic without halting the ongoing operation. This is in contrast to other custom conditions, such as a resource postcondition, which will block future resources from being applied or planned.

Conditions within assertions can refer to any scoped data source within the enclosing check block, and can refer to any variables, resources, data sources, or module outputs within the current module.

Assertions are discussed more in detail within the [Custom Conditions](/terraform/language/expressions/custom-conditions#check-assertions) documentation.

### Meta-Arguments

> In the initial release of checks, none of the traditional resource [meta-arguments](/terraform/language/resources/syntax#meta-arguments) are supported. We are interested to see how the community uses this new feature, so please file feedback with use-cases that would benefit from the introduction of any meta-arguments.

Reviewer 2: It'd be nice to have a feedback link here, do folks generally file a github issues or something else? We could add a link?

```suggestion
Check blocks do not currently support [meta-arguments](/terraform/language/resources/syntax#meta-arguments). We are still collecting feedback on this feature, so if your use case would benefit from check blocks supporting meta-arguments, please let us know. 
```

## Continuous Validation in Terraform Cloud

> Terraform Cloud can automatically check whether checks in a workspace’s configuration continue to pass after Terraform provisions the infrastructure. Refer to [Continuous Validation](/terraform/cloud-docs/workspaces/health#continuous-validation) in the Terraform Cloud documentation for details.

Reviewer 2: Avoiding using "checks" twice and cutting out a few words for brevity:
```suggestion
Terraform Cloud can automatically validate whether checks in a workspace’s configuration continue to pass after Terraform provisions new infrastructure. See [Continuous Validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for details.
```

> ## Choosing between Checks and other Custom Conditions

Reviewer 2: How do you feel about this slight shortening: 
```suggestion
## Choosing Checks or other Custom Conditions
```

    url = "https://www.terraform.io"
  }

  assert {
    condition = data.http.terraform_io.status_code == 200
    error_message = "${data.http.terraform_io.url} returned an unhealthy status code"
  }
}
```

### Scoped Data Sources

You can use any data source from any provider as a scoped data source within a `check` block.

A `check` block can optionally contain a nested (a.k.a scoped) data source. This `data` block behaves like an external [data source](/terraform/language/data-sources), except you can not reference it outside its enclosing `check` block. Additionally, if a scoped data source's provider raises any errors, they are masked as warnings and do not prevent Terraform from continuing operation execution.

You can use a scoped data source to validate the status of a piece of infrastructure outside of the usual Terraform resource lifecycle. [In the above example](#checks-syntax), if the `terraform_io` data source fails to load, you receive a warning instead of a blocking error, which would occur if you declared this data source outside of a `check` block.

Any data source from any provider can be used within a check block.

#### Meta-Arguments

> Scoped data sources support the `depends_on` and `provider` [meta-arguments](/terraform/language/resources/syntax#meta-arguments). Scoped data sources do not support the `count`, `foreach`, or `lifecycle` meta-arguments.

Reviewer 3: ```suggestion
Scoped data sources support the `depends_on` and `provider` [meta-arguments](/terraform/language/resources/syntax#meta-arguments). Scoped data sources do not support the `count`, `for_each`, or `lifecycle` meta-arguments.
```

Reviewer 3: I'm not sure if it's just typo in the docs or confusion somewhere else affecting the implementation, but `lifecycle` isn't actually supported for the `data` blocks:
https://github.com/hashicorp/terraform/blob/a801ba45cb454dea48c9ef4b1713ec1cea985594/internal/configs/resource.go#L492-L499

---
page_title: Checks - Configuration Language
description: >-
  Check customized infrastructure requirements to provide ongoing and continuous verification.
---

# Checks

-> **Note:** Check blocks are only available in Terraform v1.5.0 and later.

> The `check` block can validate your infrastructure outside the usual resource lifecycle. Check blocks address a gap between post-apply and functional validation of infrastructure. Check blocks allow you to define [custom conditions](/terraform/language/expressions/custom-conditions) that execute on every Terraform plan or apply operation without affecting the overall status of an operation. Check blocks execute as the last step of a plan or apply, after Terraform has planned or provisioned your infrastructure.

Reviewer 4: > Check blocks allow you to define [custom conditions](/terraform/language/expressions/custom-conditions) that execute on every Terraform plan or apply operation without affecting the overall status of an operation.

Forgive me, I'm still trying to work out the similarities and differences between checks and custom conditions. My understanding was that Custom Conditions™ included preconditions, postconditions, and variable validation, and we wanted to present check blocks as something distinct. 

If that's true, the language above is confusing to me. On one hand I can see how the functionality of check blocks could be described using the words "custom condition", but if checks are distinct enough to warrant their own section in the docs then I might suggest not linking out to the other feature here.

_On the other hand_, I am really wondering if it makes sense that checks land under the umbrella of Custom Conditions™, since they kind of are that. Should we consider moving this content under the Custom Condition Checks page? Does anyone else agree that might be less confusing? Am I missing any reasons we couldn't do that? Is it too late?

Reviewer 4: Please disregard the message above! It was premature. I read through the rest of the PR and the way everything fits together is making more sense to me now.

Condition arguments within `assert` blocks can refer to scoped data sources within the enclosing `check` block and any variables, resources, data sources, or module outputs within the current module.

[Learn more about assertions](/terraform/language/expressions/custom-conditions#check-assertions).

### Meta-Arguments

Check blocks do not currently support [meta-arguments](/terraform/language/resources/syntax#meta-arguments). We are still collecting feedback on this feature, so if your use case would benefit from check blocks supporting meta-arguments, please let us know. 

## Continuous Validation in Terraform Cloud

Terraform Cloud can automatically validate whether checks in a workspace’s configuration continue to pass after Terraform provisions new infrastructure. See [Continuous Validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for details.

> ## Choosing Checks or other Custom Conditions

Author: I've expanded this section a lot, I'll make a new comment thread highlighting the changes.
